emission/net/ext_service/habitica/proxy.py

```diff
@@ -3,17 +3,87 @@
 import arrow
+import urllib2
+import uuid
+import random
+
 
 # Our imports
 import emission.core.get_database as edb
 
+
 key_file = open('conf/net/keys.json')
 key_data = json.load(key_file)
 url = key_data["habitica"]["url"]
 
 
 
 def habiticaRegister(username, email, password, our_uuid):
+  user_dict = {}
+  #if user is already in e-mission db, try to load user data
+  if edb.get_habitica_db().find({'user_id': our_uuid}).count() > 0:
+    try:
+      result = habiticaProxy(our_uuid, 'GET', '/api/v3/user', None)
+      user_dict = result.json()
+      logging.debug("parsed json from GET habitica user = %s" % user_dict)
+
+    #if it fails, then user is in db but not in Habitica, so needs to create new account
+    #FIX! Still need to test if this will throw an error correctly
+    except urllib2.HTTPError:
+      user_dict = newHabiticaUser(username, email, password, our_uuid)
+
+    #now we have the user data in user_dict, so check if db is correct
+    #Fix! should prob check here if our db is right, if it's in group, etc
+
+  #if user is not in db, try to log in using email and password
+  else:
+    try:
+      login_url = url + '/api/v3/user/auth/local/login'
+      user_request = {'username': username,'email': email,'password': password}
+      logging.debug("About to login %s"% user_request)
+      u = requests.post(login_url, json=user_request)
+      user_auth = json.loads(u.text)
+      logging.debug("parsed json from habitica has keys = %s" % user_auth['data'])
+      #login only returns user auth headers, so now get authenticated user and put it in user_dict
+      auth_headers = {'x-api-user': user_auth['data']['habitica_id'], 'x-api-key': user_auth['data']['habitica_token']}
+      logging.debug("auth_headers = %s" % auth_headers['data'])
+      get_user_url = url + '/api/v3/user'
+      result = requests.request('GET', get_user_url, headers=auth_headers, json={})
+      logging.debug("result = %s" % result)
+      result.raise_for_status()
+      user_dict = result.json()
+      logging.debug("parsed json from GET habitica user = %s" % user_dict)
+
+    #if it fails, then user is also not in Habitica, so needs to create new account and put it in user_dict
+    #FIX!! throw except only if u returns a 401 error
+    except:
+      user_dict = newHabiticaUser(username, email, password, our_uuid)
+      
+    #Now save new user (user_dict) to our db
+    #Since we are randomly generating the password, we store it in case users 
+    #want to access their Habitica account from the browser
+    #Need to create a way from them to retrieve username/password
+    #metrics_data is used to calculate points based on km biked/walked
+    #last_timestamp is the last time the user got points, and bike/walk_count are the leftover km
+    habitica_user_table = edb.get_habitica_db()
+    habitica_user_table.insert({'user_id': our_uuid, 
+      'metrics_data': {'last_timestamp': arrow.utcnow().timestamp, 'bike_count': 0, 'walk_count': 0},
```

Comment:
shouldn't you create the user only if she wasn't in the emission database?
And you should update if the user was in emission but not in habitica. Otherwise, you will end up with two entries for the same user, right?